python/ql/src/experimental/Security/CWE-022bis/UnsafeUnpack.ql

```diff
@@ -0,0 +1,136 @@
+ * @problem.severity error
+ * @security-severity 7.5
+ * @precision high
+ * @tags security
+ *       external/cwe/cwe-022
+ */
+
+import python
+import semmle.python.Concepts
+import semmle.python.dataflow.new.internal.DataFlowPublic
+import semmle.python.ApiGraphs
+import DataFlow::PathGraph
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.frameworks.Stdlib
+
+class UnsafeUnpackingConfig extends TaintTracking::Configuration {
+  UnsafeUnpackingConfig() { this = "UnsafeUnpackingConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    // A source coming from a remote location
+    exists(Http::Client::Request request | source = request)
+    or
+    //A source coming from a CLI argparse module
+    exists(Node o, API::Node ap, MethodCallNode args |
+      ap = API::moduleImport("argparse").getMember("ArgumentParser").getACall().getReturn() and
+      args = ap.getMember("parse_args").getACall() and
+      args.flowsTo(o) and
+      source.(AttrRead).accesses(o, any(string s))
+    )
+    or
+    // A source catching a S3 filename download
+    exists(API::Node s3 | source = s3.getMember("download_file").getACall().getArg(2))
+    or
+    // A source download a file using wget
+    exists(MethodCallNode mcn |
+      mcn = API::moduleImport("wget").getMember("download").getACall() and source = mcn.getArg(1)
+    )
+    or
+    // catch the uploaded files as a source
+    exists(Subscript s, Attribute at |
+      at = s.getObject() and at.getAttr() = "FILES" and source.asExpr() = s
+    )
+    or
+    exists(Node obj, AttrRead ar |
+      ar.getAMethodCall("get").flowsTo(source) and
+      ar.accesses(obj, "FILES")
+    )
+    or
+    exists(Node obj, AttrRead ar |
+      ar.getAMethodCall("getlist").flowsTo(source) and
+      ar.accesses(obj, "FILES")
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    // A sink capturing method calls to `unpack_archive`.
+    sink = API::moduleImport("shutil").getMember("unpack_archive").getACall().getArg(0)
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    // Writing the response data to the archive
+    exists(Stdlib::FileLikeObject::InstanceSource is, Node f, MethodCallNode mc |
+      is.flowsTo(f) and
+      mc.getMethodName() = "write" and
+      f = mc.getObject() and
+      nodeFrom = mc.getArg(0) and
+      nodeTo = is.(CallCfgNode).getArg(0)
+    )
+    or
+    // Copying the response data to the archive
+    exists(Stdlib::FileLikeObject::InstanceSource is, Node f, MethodCallNode mc |
+      is.flowsTo(f) and
+      mc = API::moduleImport("shutil").getMember("copyfileobj").getACall() and
+      f = mc.getArg(1) and
+      nodeFrom = mc.getArg(0) and
+      nodeTo = is.(CallCfgNode).getArg(0)
+    )
+    or
+    // Reading the response
+    exists(MethodCallNode mc |
+      nodeFrom = mc.getObject() and
+      mc.getMethodName() = "read" and
+      mc.flowsTo(nodeTo)
```

Comment:
(Here and other places where you write `.flowsTo(nodeTo)`)
There is no need to track flow further, since we are adding a step to the flow graph. Otherwise, we will just be adding many big steps. We should just add the one small step and let data flow handle the rest.
```suggestion
      nodeTo = mc
```

Reply:
I fixed this one. I did not need a `.flowsTo(nodeTo)` for the `read()` for sure.

Reply:
The current version of query uses `.flowsTo()` twice in:

```codeql
    // Open a file for access
    exists(MethodCallNode cn |
      cn.calls(nodeTo, "open") and
      cn.flowsTo(nodeFrom)
    )
    or
    // Open a file for access using builtin
    exists(API::CallNode cn |
      cn = API::builtin("open").getACall() and
      nodeTo = cn.(API::CallNode).getArg(0) and
      cn.flowsTo(nodeFrom)
    )
```

This is due to the potential use of context managers or not. I mean either `f=open() and close(f)` or `with open(f) as f1: ...`.